### DISCUSSION TITLE
Reasons for using 'android_event_quque'?

### DISCUSSION THREAD

**Discussion by kimcoldwater** [FreeRDP/FreeRDP#8115] [Q&A] [answered]
hello,

I am making RDP available on the web using JNI and FreeRDP like the Android version.

In Android_client,

If look at the `Android_freerdp_run` function in `Android_freerdp.c`, I can see that there is an `Android_check_handle` function in the main loop.

I know the structure and use of it.

However, when I call events such as keyboard and mouse using jni, I call something like 'freeerdp_client_send_button_event' directly within the JNI function without using a queue.

In short, event calls -> insert into quque -> main loop checks the queue and does not go through the logic of event processing.

Is there any problem if I call the event immediately without using the main loop?

Because I use JNI, the stability of the freeRDP library is very important (the server is forced to shut down in case of a crash).

Can you tell me why use this method?

**Comment by mfleisz** [comment, marked as answer]
@kimcoldwater 

As far as I can remember that code was written at a time where the freerdp send input event functions were not thread-safe.

With the current library it should be no problem to call these functions directly from any thread.
